CASSANDRA-16575: Refactor unit test configurations

jacek-lewandowski opened the pull request:
Instead of predefined test, test-compression, test-cdc, test-system-keyspace-directory, have the ability to specify a properties file, which may include extra-java-opts property. That property contains extra JVM options to be added to a forked JVM for testing.

This way, instead of:

ant test-compression

we have:
ant -Dprops=test/conf/unit-with-sstable-compression.properties test

We can easily create new configurations or modify existing without touching the build itself

Also changes the compressor used in compression configuration from snappy to lz4

michaelsembwever commented on build.xml:
```diff
@@ -1444,75 +1447,6 @@
-      <property name="cdc_yaml" value="${build.test.dir}/cassandra.cdc.yaml"/>
-      <concat destfile="${cdc_yaml}">
-        <fileset file="${test.conf}/cassandra.yaml"/>
-        <fileset file="${test.conf}/cdc.yaml"/>
```
is there any way of maintaining this feature?  it is cumbersome (and error-prone) to keep all the different test cassandra.yaml's up to date…

michaelsembwever commented on build.xml:
```diff
@@ -69,6 +69,8 @@
     <property name="dist.dir" value="${build.dir}/dist"/>
     <property name="tmp.dir" value="${java.io.tmpdir}"/>
+    <property file="${props}"/>
+    <property name="extra-java-opts" value=""/>
```
would it be possible to encapsulate these just related test targets?

michaelsembwever commented on test/conf/commitlog_compression_Zstd.yaml:
```diff
@@ -1,2 +0,0 @@
-commitlog_compression:
-    - class_name: ZstdCompressor
```
this should be maintained. compression tests should enable both sstable and commitlog compression.

michaelsembwever commented:
A few small comments added.

Compatibility over branches is also a concern here. We currently have one [build script](https://github.com/apache/cassandra-builds/blob/trunk/build-scripts/cassandra-test.sh#L93-L104) that kicks off the ant command line for the different test targets, that runs regardless of the cassandra branch.

Commit:
CASSANDRA-16575: Fix compression tests

By default we will use LZ4, which is the default in C*. Also we will take appropriate commitlog_compression_xxx.yaml file, adequate to the selected algorithm. the algorithm is defined in test.compression_algo, the property which existed before but was unused.

michaelsembwever commented on test/unit/org/apache/cassandra/SchemaLoader.java:
```diff
@@ -751,4 +747,26 @@ public static void cleanupSavedCaches()
+
+    private static CompressionParams compressionParams(int chunkLength)
+    {
+        String algo = System.getProperty("test.compression_algo", "lz4").toLowerCase();
```
```suggestion
        String algo = System.getProperty("cassandra.test.compression.algo", "lz4").toLowerCase();
```

michaelsembwever reviewed (approved):
tested with 
```
ant test-compression -Dtest.name=BigTableWriterTest
ant test-compression -Dtest.name=BigTableWriterTest -Dtest.compression.algo=lz4
ant test-compression -Dtest.name=BigTableWriterTest -Dtest.compression.algo=snappy
ant test-compression -Dtest.name=BigTableWriterTest -Dtest.compression.algo=noop
ant test-compression -Dtest.name=BigTableWriterTest -Dtest.compression.algo=zstd
ant test-compression -Dtest.name=BigTableWriterTest -Dtest.compression.algo=none
ant test-compression -Dtest.name=BigTableWriterTest -Dtest.compression.algo=jjj

```

michaelsembwever commented:
merged with https://github.com/apache/cassandra/commit/10a1d65eb09a93aee32948b46b4f1a0fbc2defe0
